external-crates/move/crates/move-vm-runtime/src/loader/package_loader.rs

```diff
@@ -0,0 +1,358 @@
+            if module.address() != &runtime_package_id {
+                return Err(verification_error(
+                    StatusCode::MISMATCHED_MODULE_IDS_IN_PACKAGE,
+                    IndexKind::AddressIdentifier,
+                    module.self_handle_idx().0,
+                )
+                .finish(Location::Undefined));
+            }
+        }
+
+        let storage_id = {
+            let module_id = loading_package
+                .as_modules()
+                .into_iter()
+                .next()
+                .expect("non-empty package")
+                .self_id();
+            *data_store
+                .relocate(&module_id)
+                .map_err(|e| e.finish(Location::Undefined))?
+                .address()
+        };
+
+        // Cache the package's dependencies without the package.
+        let cached_packages =
+            self.load_and_cache_packages(data_store, data_store.all_package_dependencies()?)?;
+
+        // Now verify linking on-the-spot to make sure that the current package links correctly in
+        // the supplied linking context.
+        linkage_checker::verify_linkage_and_cyclic_checks_for_publication(
+            &loading_package,
+            &cached_packages,
+        )?;
+
+        // Cache the package and its types.
+        self.package_cache.write().cache_package(
+            storage_id,
+            loading_package,
+            &self.natives,
+            data_store,
+            &self.type_cache,
+        )?;
+
+        Ok(())
+    }
+
+    // Loads the set of packages into the package cache.
+    fn load_and_cache_packages(
+        &self,
+        data_store: &impl DataStore,
+        packages_to_read: BTreeSet<PackageStorageId>,
+    ) -> VMResult<BTreeMap<PackageStorageId, Arc<LoadedPackage>>> {
+        let allow_loading_failure = true;
+        let root_package = data_store.link_context();
+
+        let mut dependency_order = vec![];
+        let mut seen_packages = BTreeSet::new();
+
+        let mut cached_packages = BTreeMap::new();
+        let mut pkgs_to_cache = BTreeMap::new();
+        let mut work_queue = vec![root_package];
+
+        // Load all packages, compute dependency order (including possibly already cached
+        // packages). NB: packages can be loaded out of order here if so desired.
+        while let Some(dep) = work_queue.pop() {
+            if seen_packages.contains(&dep) {
+                continue;
+            }
+
+            // Check if package is already cached. If so add it to the cached packages.
+            // Also compute the packages dependency order. This is because we need to count on the fact that
+            // all dependencies are loaded and their types cached before we cache a package.
+            let package_deps = if let Some(pkg) = self.package_cache.read().loaded_package_at(dep) {
+                let package_deps = Self::compute_immediate_package_dependencies(
```

Comment:
Don't we have a guarantee that the dependencies are in the cache if we find this package in the cache?

Reply:
Not necessarily since different link contexts will induce different dependencies possibly and a package is not cached based on link context (and is in fact shared across multiple link contexts).